### PR TITLE
Unify / clean-up ADC driver

### PR DESCRIPTION
**Pull request by raphaelcoeffic**
## Description
This an attempt at re-writing the ADC driver to be cleaner than the existing.

It bases on 2 layers of abstraction:
- generic ADC driver (see `hal/adc_driver.h`)
- generic STM32 ADC driver (see `targets/common/arm/stm32/stm32_hal_adc.h`)

It has been tested on:
- [x] T-LITE (taranis / STM32F2)
- [x] T16 (horus / STM32F4)
- [x] NV14 (horus / STM32F4)
- [x] X9-lite (taranis / STM32F2) 
- [x] X12S (horus / STM32F4; external SPI ADC)

## Generic ADC driver

This driver offers only basic methods:
```
struct etx_hal_adc_driver_t {
  void (*init)();
  void (*start_conversion)();
  void (*wait_completion)();  
};

void adcInit(const etx_hal_adc_driver_t* driver);
void     adcRead();
uint16_t getRTCBatteryVoltage();
uint16_t getAnalogValue(uint8_t index);
```

TODO:
- [ ] adding mapping & direction table directly to the driver structure.

## Generic STM32 ADC driver

This driver supports DMA based multi-channel reads. Once the channels are setup, the conversion is started by issuing a DMA request (for those ADC definitions that supports it) or triggering a single conversion for ADC definitions (no DMA).

Please note that this driver is heavily based on the StdPeriph library and needs to be converted to Low-Layer HAL.

TODO:
- [ ] improve wait method (hardcoded for now)
- [ ] improve DMA stop (hardcoded for now)
- [ ] add support for completion IRQ (DMA and single conversions)

### CODE REVIEW

File: radio/src/targets/common/arm/stm32/stm32_hal_adc.cpp [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+  ADC_Init(ADCx, &ADC_InitStructure);
+
+  // enable ADC_MAIN
+  ADC_Cmd(ADCx, ENABLE);                            // ADC_CR2_ADON
+  // enable DMA for ADC
+  ADC_DMACmd(ADCx, ENABLE);                         // ADC_CR2_DMA
+  ADC_DMARequestAfterLastTransferCmd(ADCx, ENABLE); // ADC_CR2_DDS
+}
+
+stm32_hal_adc_channel ADC_MAIN_channels[] = {
```
Comment (raphaelcoeffic) [inline radio/src/targets/common/arm/stm32/stm32_hal_adc.cpp:74]: Channel definitions need to move to the board specific files.

**Comment by rotorman** [comment]
I guess this line needs to be removed:
https://github.com/EdgeTX/edgetx/blob/8add1515491cd6218ffb260fd1b28181da07ecf1/radio/src/targets/common/arm/stm32/stm32_hal_adc.h#L26

**Comment by raphaelcoeffic** [comment]
> I guess this line needs to be removed:
> 
> https://github.com/EdgeTX/edgetx/blob/8add1515491cd6218ffb260fd1b28181da07ecf1/radio/src/targets/common/arm/stm32/stm32_hal_adc.h#L26
> 
> as otherwise this definition would need 3 elements each in define, but as the rank is dynamically given in _adc_init_channels()_, it would not make sense to list here:
> 
> https://github.com/EdgeTX/edgetx/blob/8add1515491cd6218ffb260fd1b28181da07ecf1/radio/src/targets/common/arm/stm32/stm32_hal_adc.cpp#L74-L114
> 
> 
> Currently the sampling time is not defined (would be third element that is not given).

Yes, that‘s right! Stupid last minute mistake! Thx!

**Comment by TramALot** [comment]
I been watching the jitter change for 2 weeks 5 different builds and the
edge-tx-firmware-main 2021-07-27T15:47:15z has zero on my t18 radioking.

The outputs widget is locked on 0 for AER channels at center like the
Throttle has been!

ADC filter on.

Thanks for all of all of your works

**Pull request merged**